src/daemon/rpmostree-sysroot-upgrader.c

```diff
@@ -561,13 +562,60 @@ checkout_base_tree (RpmOstreeSysrootUpgrader *self,
                        &self->tmprootfs_dfd, error))
     return FALSE;
 
-  /* build a centralized rsack for it, since we need it in a few places */
-  self->rsack = rpmostree_get_refsack_for_root (self->tmprootfs_dfd, ".", error);
+  rpmostree_output_task_end ("done");
+  return TRUE;
+}
+
+static gboolean
+load_base_rsack (RpmOstreeSysrootUpgrader *self,
+                 GCancellable             *cancellable,
+                 GError                  **error)
+{
+  /* go through all the deployments and check if one of them has the base rev */
+  g_autoptr(GPtrArray) deployments = ostree_sysroot_get_deployments (self->sysroot);
+  for (guint i = 0; i < deployments->len; i++)
+    {
+      OstreeDeployment *deployment = deployments->pdata[i];
+
+      gboolean is_layered;
+      g_autofree char *base_rev_owned = NULL;
+      if (!rpmostree_deployment_get_layered_info (self->repo, deployment,
+                                                  &is_layered, &base_rev_owned,
+                                                  NULL, NULL, NULL, error))
+        return FALSE;
+
+      const char *base_rev = base_rev_owned ?: ostree_deployment_get_csum (deployment);
+      if (g_str_equal (self->base_revision, base_rev))
+        {
+          int sysroot_fd = ostree_sysroot_get_fd (self->sysroot);
+          g_autofree char *path =
+            ostree_sysroot_get_deployment_dirpath (self->sysroot, deployment);
+
+          if (is_layered)
+            self->rsack = rpmostree_get_base_refsack_for_root (sysroot_fd, path,
```

Comment:
Oh nevermind, I see we do here.  Why conditionalize on `is_layered`?  Can't we just unconditionally call `get_base_refsack()` and it should Just Work in both cases?

I guess I'm also confused by this "searching all deployments" thing - the SysrootUpgrader operates off of the pending deployment by default, shouldn't we grab the base rpmdb just from that deployment?

Reply:
>  Why conditionalize on `is_layered`? Can't we just unconditionally call `get_base_refsack()` and it should Just Work in both cases?

That just moves the logic from here to `rpmostree-rpm-util.h`, right? I.e. instead of `if (is_layered)` here, we'd check for `RPMOSTREE_BASE_RPMDB` there and fall back to `RPMOSTREE_RPMDB_LOCATION`. I find keeping the logic here keeps things clean, esp. since we have to call `rpmostree_deployment_get_layered_info` anyway to get the base layer, so we get `is_layered` "for free".

> the SysrootUpgrader operates off of the pending deployment by default, shouldn't we grab the base rpmdb just from that deployment?

Yeah, that's the common case, in which we'd stop at the first iteration. I guess I was also generalizing for the cases where we're targeting a base layer that's not the pending deployment? (E.g. doing `deploy $VERSION` where `$VERSION` is currently the rollback). Basically, it's really cheap to opportunistically check for pre-existing checkouts compared to doing a full checkout.

Reply:
Ahh nice, I had missed that.
Yeah, that's definitely better then.